File: framec_tests/python/src/state_stack/state_stack.py
# Emitted from framec_v0.11.0



from framelang.framelang import FrameEvent



class FrameEvent:
    def __init__(self, message, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class StateStack:
    
    
    # ==================== System Factory =================== #
    
    def __init__(self):
        
        # Create state stack.
        
        self.__state_stack = []
        
         # Create and intialize start state compartment.
        
        self.__compartment: 'StateStackCompartment' = StateStackCompartment('__statestack_state_A')
        self.__next_compartment: 'StateStackCompartment' = None
        
        # Initialize domain
        
        self.tape  = []
        
        # Send system start event
        frame_event = FrameEvent(">", None)
        self.__kernel(frame_event)
    
    # ==================== Interface Block ================== #
    
    def to_a(self,):
        __e = FrameEvent("to_a",None)
        self.__kernel(__e)
    
    def to_b(self,):
        __e = FrameEvent("to_b",None)
        self.__kernel(__e)
    
    def to_c(self,):
        __e = FrameEvent("to_c",None)
        self.__kernel(__e)
    
    def push(self,):
        __e = FrameEvent("push",None)
        self.__kernel(__e)
    
    def pop(self,):
        __e = FrameEvent("pop",None)
        self.__kernel(__e)
    
    # ===================== Machine Block =================== #
    
    # ----------------------------------------
    # $A
    
    def __statestack_state_A(self, __e):
        if __e._message == ">":
            self.log_do("A:>")
            return
        elif __e._message == "<":
            self.log_do("A:<")
            return
        elif __e._message == "to_a":
            next_compartment = StateStackCompartment('__statestack_state_A')
            self.__transition(next_compartment)
            return
        elif __e._message == "to_b":
            next_compartment = StateStackCompartment('__statestack_state_B')
            self.__transition(next_compartment)
            return
        elif __e._message == "to_c":
            next_compartment = StateStackCompartment('__statestack_state_C')
            self.__transition(next_compartment)
            return
        elif __e._message == "push":
            self.__state_stack_push(self.__compartment)
            return
        elif __e._message == "pop":
            next_compartment = self.__state_stack_pop()
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $B
    
    def __statestack_state_B(self, __e):
        if __e._message == ">":
            self.log_do("B:>")
            return
        elif __e._message == "<":
            self.log_do("B:<")
            return
        elif __e._message == "to_a":
            next_compartment = StateStackCompartment('__statestack_state_A')
            self.__transition(next_compartment)
            return
        elif __e._message == "to_b":
            next_compartment = StateStackCompartment('__statestack_state_B')
            self.__transition(next_compartment)
            return
        elif __e._message == "to_c":
            next_compartment = StateStackCompartment('__statestack_state_C')
            self.__transition(next_compartment)
            return
        elif __e._message == "push":
            self.__state_stack_push(self.__compartment)
            return
        elif __e._message == "pop":
            next_compartment = self.__state_stack_pop()
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $C
    
    def __statestack_state_C(self, __e):
        if __e._message == ">":
            self.log_do("C:>")
            return
        elif __e._message == "<":
            self.log_do("C:<")
            return
        elif __e._message == "to_a":
            next_compartment = StateStackCompartment('__statestack_state_A')
            self.__transition(next_compartment)
            return
        elif __e._message == "to_b":
            next_compartment = StateStackCompartment('__statestack_state_B')
            self.__transition(next_compartment)
            return
        elif __e._message == "to_c":
            next_compartment = StateStackCompartment('__statestack_state_C')
            self.__transition(next_compartment)
            return
        elif __e._message == "push":
            self.__state_stack_push(self.__compartment)
            return
        elif __e._message == "pop":
            next_compartment = self.__state_stack_pop()
            self.__transition(next_compartment)
            return
    
    # ===================== Actions Block =================== #
    
    def log_do(self,msg: str):
        raise NotImplementedError
    
    # ==================== System Runtime =================== #
    
    def __kernel(self, __e):
        
        # send event to current state
        self.__router(__e)
        
        # loop until no transitions occur
        while self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            
            # exit current state
            self.__router(FrameEvent( "<", self.__compartment.exit_args))
            # change state
            self.__compartment = next_compartment
            
            if next_compartment.forward_event is None:
                # send normal enter event
                self.__router(FrameEvent(">", self.__compartment.enter_args))
            else: # there is a forwarded event
                if next_compartment.forward_event._message == ">":
                    # forwarded event is enter event
                    self.__router(next_compartment.forward_event)
                else:
                    # forwarded event is not enter event
                    # send normal enter event
                    self.__router(FrameEvent(">", self.__compartment.enter_args))
                    # and now forward event to new, intialized state
                    self.__router(next_compartment.forward_event)
                next_compartment.forward_event = None
                
    
    def __router(self, __e):
        if self.__compartment.state == '__statestack_state_A':
            self.__statestack_state_A(__e)
        elif self.__compartment.state == '__statestack_state_B':
            self.__statestack_state_B(__e)
        elif self.__compartment.state == '__statestack_state_C':
            self.__statestack_state_C(__e)
        
    def __transition(self, next_compartment: 'StateStackCompartment'):
        self.__next_compartment = next_compartment
    
    def __state_stack_push(self, compartment: 'StateStackCompartment'):
        self.__state_stack.append(compartment)
    
    def __state_stack_pop(self):
        return self.__state_stack.pop()
    
    def state_info(self):
        return self.__compartment.state
        

# ===================== Compartment =================== #

class StateStackCompartment:

    def __init__(self,state):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
    

File: framec_tests/golang/state_stack/state_stack.go
// emitted from framec_v0.11.0
// get include files at https://github.com/frame-lang/frame-ancillary-files


func NewStateStack() StateStack {
    m := &stateStackStruct{}
    
    // Validate interfaces
    var _ StateStack = m
    var _ StateStack_actions = m
    // History mechanism used in spec. Create state stack.
    m._stateStack_ = &Stack{stack: make([]StateStackCompartment, 0)}
    
    
    // Create and intialize start state compartment.
    m._compartment_ = NewStateStackCompartment(StateStackState_A)
    
    // Override domain variables.
    m.tape = []string{}
    
    // Send system start event
    e := framelang.FrameEvent{Msg:">"}
    m._mux_(&e)
    return m
}

type Stack struct {
    stack []StateStackCompartment
}

func (s *Stack) Push(compartment *StateStackCompartment) {
    s.stack = append(s.stack, *compartment)
}

func (s *Stack) Pop() *StateStackCompartment {

    l := len(s.stack)
    if l == 0 {
        panic("Attempted to pop when history stack is empty")
    }
    
    res := s.stack[l-1]
    s.stack = s.stack[:l-1]
    return &res
}

type StateStackState uint

const (
    StateStackState_A StateStackState = iota
    StateStackState_B
    StateStackState_C
)

type StateStack interface {
    To_a() 
    To_b() 
    To_c() 
    Push() 
    Pop() 
    Pop_change() 
}

type StateStack_actions interface {
    log(msg string) 
}


type stateStackStruct struct {
    _compartment_ *StateStackCompartment
    _nextCompartment_ *StateStackCompartment
    _stateStack_ *Stack
    tape []string
}

//===================== Interface Block ===================//

func (m *stateStackStruct) To_a()  {
    e := framelang.FrameEvent{Msg:"to_a"}
    m._mux_(&e)
}

func (m *stateStackStruct) To_b()  {
    e := framelang.FrameEvent{Msg:"to_b"}
    m._mux_(&e)
}

func (m *stateStackStruct) To_c()  {
    e := framelang.FrameEvent{Msg:"to_c"}
    m._mux_(&e)
}

func (m *stateStackStruct) Push()  {
    e := framelang.FrameEvent{Msg:"push"}
    m._mux_(&e)
}

func (m *stateStackStruct) Pop()  {
    e := framelang.FrameEvent{Msg:"pop"}
    m._mux_(&e)
}

func (m *stateStackStruct) Pop_change()  {
    e := framelang.FrameEvent{Msg:"pop_change"}
    m._mux_(&e)
}

//====================== Multiplexer ====================//

func (m *stateStackStruct) _mux_(e *framelang.FrameEvent) {
    switch m._compartment_.State {
    case StateStackState_A:
        m._StateStackState_A_(e)
    case StateStackState_B:
        m._StateStackState_B_(e)
    case StateStackState_C:
        m._StateStackState_C_(e)
    }
    
    if m._nextCompartment_ != nil {
        nextCompartment := m._nextCompartment_
        m._nextCompartment_ = nil
        if nextCompartment._forwardEvent_ != nil && 
           nextCompartment._forwardEvent_.Msg == ">" {
            m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
            m._compartment_ = nextCompartment
            m._mux_(nextCompartment._forwardEvent_)
        } else {
            m._do_transition_(nextCompartment)
            if nextCompartment._forwardEvent_ != nil {
                m._mux_(nextCompartment._forwardEvent_)
            }
        }
        nextCompartment._forwardEvent_ = nil
    }
}

//===================== Machine Block ===================//

func (m *stateStackStruct) _StateStackState_A_(e *framelang.FrameEvent) {
    switch e.Msg {
    case ">":
        m.log("A:>")
        
        return
    case "<":
        m.log("A:<")
        
        return
    case "to_a":
        compartment := NewStateStackCompartment(StateStackState_A)
        m._transition_(compartment)
        
        return
    case "to_b":
        compartment := NewStateStackCompartment(StateStackState_B)
        m._transition_(compartment)
        
        return
    case "to_c":
        compartment := NewStateStackCompartment(StateStackState_C)
        m._transition_(compartment)
        
        return
    case "push":
        m._stateStack_push_(m._compartment_)
        
        return
    case "pop":
        compartment := m._stateStack_pop_()
        m._transition_(compartment)
        
        return
    case "pop_change":
        
        return
    }
}  // ->> $$[-]


func (m *stateStackStruct) _StateStackState_B_(e *framelang.FrameEvent) {
    switch e.Msg {
    case ">":
        m.log("B:>")
        
        return
    case "<":
        m.log("B:<")
        
        return
    case "to_a":
        compartment := NewStateStackCompartment(StateStackState_A)
        m._transition_(compartment)
        
        return
    case "to_b":
        compartment := NewStateStackCompartment(StateStackState_B)
        m._transition_(compartment)
        
        return
    case "to_c":
        compartment := NewStateStackCompartment(StateStackState_C)
        m._transition_(compartment)
        
        return
    case "push":
        m._stateStack_push_(m._compartment_)
        
        return
    case "pop":
        compartment := m._stateStack_pop_()
        m._transition_(compartment)
        
        return
    case "pop_change":
        
        return
    }
}  // ->> $$[-]


func (m *stateStackStruct) _StateStackState_C_(e *framelang.FrameEvent) {
    switch e.Msg {
    case ">":
        m.log("C:>")
        
        return
    case "<":
        m.log("C:<")
        
        return
    case "to_a":
        compartment := NewStateStackCompartment(StateStackState_A)
        m._transition_(compartment)
        
        return
    case "to_b":
        compartment := NewStateStackCompartment(StateStackState_B)
        m._transition_(compartment)
        
        return
    case "to_c":
        compartment := NewStateStackCompartment(StateStackState_C)
        m._transition_(compartment)
        
        return
    case "push":
        m._stateStack_push_(m._compartment_)
        
        return
    case "pop":
        compartment := m._stateStack_pop_()
        m._transition_(compartment)
        
        return
    case "pop_change":
        
        return
    }
}

//=============== Machinery and Mechanisms ==============//

func (m *stateStackStruct) _transition_(compartment *StateStackCompartment) {
    m._nextCompartment_ = compartment
}

func (m *stateStackStruct) _do_transition_(nextCompartment *StateStackCompartment) {
    m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
    m._compartment_ = nextCompartment
    m._mux_(&framelang.FrameEvent{Msg: ">", Params: m._compartment_.EnterArgs, Ret: nil})
}

func (m *stateStackStruct) _stateStack_push_(compartment *StateStackCompartment) {
    m._stateStack_.Push(compartment)
}

func (m *stateStackStruct) _stateStack_pop_() *StateStackCompartment {
    compartment := m._stateStack_.Pop()
    return compartment
}

//===================== Actions Block ===================//


/********************************************************

// Unimplemented Actions

func (m *stateStackStruct) log(msg string)  {}

********************************************************/

//=============== Compartment ==============//

type StateStackCompartment struct {
    State StateStackState
    StateArgs map[string]interface{}
    StateVars map[string]interface{}
    EnterArgs map[string]interface{}
    ExitArgs map[string]interface{}
    _forwardEvent_ *framelang.FrameEvent
}

func NewStateStackCompartment(state StateStackState) *StateStackCompartment {
    c := &StateStackCompartment{State: state}
    c.StateArgs = make(map[string]interface{})
    c.StateVars = make(map[string]interface{})
    c.EnterArgs = make(map[string]interface{})
    c.ExitArgs = make(map[string]interface{})
    return c
}